File: cloudflare-backend/src/routes/transactions.ts
// src/routes/transactions.ts
import { Hono } from 'hono';
import { AppEnv, getCurrentUser } from '../middleware/auth';
import { Permission, hasPermission } from '../permissions';
import { DbCard, DbTransaction, DbUser, TransactionType } from '../types';
import { generateUuid } from '../utils';

const transactions = new Hono<AppEnv>();

transactions.post('/', async (c) => {
  const actor = (await getCurrentUser(c)) || c.get('user');
  if (!actor) {
    return c.json({ detail: 'Sign in to continue.' }, 401);
  }

  const body = await c.req.json<{
    card_uid: string;
    amount: number;
    transaction_type: TransactionType;
    description?: string | null;
  }>();

  if (!body.card_uid?.trim() || body.amount === undefined || !body.transaction_type) {
    return c.json({ detail: 'card_uid, amount and transaction_type are required' }, 400);
  }

  // Authorize direction
  const neededPermission =
    body.transaction_type === TransactionType.RECHARGE
      ? Permission.TOP_UP_WALLET
      : Permission.CHARGE_TOKENS;

  if (!hasPermission(actor.role, neededPermission)) {
    const detail =
      body.transaction_type === TransactionType.RECHARGE
        ? 'Only an admin can top up a wallet.'
        : "Your role can't take payment.";
    return c.json({ detail }, 403);
  }

  // Resolve user
  const card = await c.env.DB.prepare('SELECT * FROM cards WHERE uid = ?')
    .bind(body.card_uid.trim())
    .first<DbCard>();

  if (!card) {
    return c.json({ detail: 'Card not registered' }, 404);
  }

  const user = await c.env.DB.prepare('SELECT * FROM users WHERE id = ?')
    .bind(card.user_id)
    .first<DbUser>();

  if (!user) {
    return c.json({ detail: 'User not found' }, 404);
  }

  let newBalance = user.token_balance;
  if (body.transaction_type === TransactionType.RECHARGE) {
    newBalance += body.amount;
  } else if (body.transaction_type === TransactionType.DEDUCTION) {
    if (user.token_balance < body.amount) {
      return c.json({ detail: 'Insufficient balance' }, 400);
    }
    newBalance -= body.amount;
  }

  const txId = generateUuid();
  const createdAt = new Date().toISOString();

  // Execute update balance and record transaction atomically in D1
  await c.env.DB.batch([
    c.env.DB.prepare('UPDATE users SET token_balance = ? WHERE id = ?').bind(newBalance, user.id),
    c.env.DB.prepare(
      'INSERT INTO transactions (id, user_id, amount, transaction_type, description, created_at) VALUES (?, ?, ?, ?, ?, ?)'
    ).bind(
      txId,
      user.id,
      body.amount,
      body.transaction_type,
      body.description || null,
      createdAt
    ),
  ]);

  const createdTx = await c.env.DB.prepare('SELECT * FROM transactions WHERE id = ?')
    .bind(txId)
    .first<DbTransaction>();

  if (!createdTx) {
    return c.json({ detail: 'Failed to record transaction' }, 500);
  }

  return c.json(
    {
      id: createdTx.id,
      user_id: createdTx.user_id,
      amount: createdTx.amount,
      transaction_type: createdTx.transaction_type as TransactionType,
      description: createdTx.description,
      created_at: createdTx.created_at,
    },
    201
  );
});

export default transactions;

File: backend/app/routers/transactions.py
from fastapi import APIRouter, HTTPException, status, Depends
from sqlalchemy.orm import Session
from datetime import datetime

from app.auth_deps import current_user
from app.permissions import Permission, has_permission
from app.schemas import TransactionRequest, TransactionResponse, TransactionType
from app.db.database import get_db
from app.models.models import User, Card, TransactionRecord

router = APIRouter(prefix="/transactions", tags=["transactions"])

# Which permission each direction of money movement requires. Crediting a wallet and debiting it are
# separate rights: an organizer at the till has CHARGE_TOKENS but not TOP_UP_WALLET.
_REQUIRED_PERMISSION = {
    TransactionType.recharge: Permission.TOP_UP_WALLET,
    TransactionType.deduction: Permission.CHARGE_TOKENS,
}


@router.post("", response_model=TransactionResponse, status_code=status.HTTP_201_CREATED)
def process_transaction(
    payload: TransactionRequest,
    actor: User = Depends(current_user),
    db: Session = Depends(get_db),
) -> TransactionResponse:
    # 0. Authorize the specific direction. Checked before touching any balance.
    needed = _REQUIRED_PERMISSION[payload.transaction_type]
    if not has_permission(actor.role, needed):
        detail = (
            "Only an admin can top up a wallet."
            if payload.transaction_type == TransactionType.recharge
            else "Your role can't take payment."
        )
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=detail)

    # 1. Resolve User
    card = db.query(Card).filter(Card.uid == payload.card_uid).first()
    if not card:
        raise HTTPException(status_code=404, detail="Card not registered")
    
    user = card.user
    
    # 2. Process based on type
    if payload.transaction_type == TransactionType.recharge:
        user.token_balance += payload.amount
    elif payload.transaction_type == TransactionType.deduction:
        if user.token_balance < payload.amount:
            raise HTTPException(status_code=400, detail="Insufficient balance")
        user.token_balance -= payload.amount
        
    # 3. Record Transaction
    transaction = TransactionRecord(
        user_id=user.id,
        amount=payload.amount,
        transaction_type=payload.transaction_type,
        description=payload.description,
        created_at=datetime.utcnow()
    )
    
    db.add(transaction)
    db.add(user) # Update user balance
    db.commit()
    db.refresh(transaction)
    
    return transaction
